docs/proposals/image-exec.md

```diff
@@ -0,0 +1,278 @@
+run inside the pod namespace.
+
+The user will specify a debug image of her choice (e.g. `kubectl exec -it -m
+gcr.io/myproj/debugimage podname`) and be connected to a process running inside
+the pod. The processes of all other containers would be visible through the
+shared PID namespace (if enabled). Network connections are visible through the
+shared net namespace. The kubelet will bind mount the filesystems of other
+containers in the pod into the debug container (e.g. in `/c/<container_name>/`),
+potentially read-only.
+
+The lifecycle of the debug container will model that of the current `kubectl
+exec`, with the container being cleaned up after it exits. Since docker doesn't
+support exec in a container that hasn't started, this is implemented under the
+covers as create, start, attach, and remove in immediate succession.
+
+## API Changes
+
+The Exec handler will gain a new parameter to specify a debug image.
+
+TODO: details of API changes
+
+## Changes to kubelet
+
+The kubelet will gain a new `ExecInPod` method to be called by `ServeExec`
+instead of `ExecInContainer` when an container image name is provided. Image
+exec could be accommodated by the current `ExecInContainer` with the addition of
+a parameter for imageName, but it would the method name misleading.
+
+A corresponding `ExecInPod` will be added to the kubelet's container.Runtime
+interface.
+
+### Container Runtime Interface
+
+The [Container Runtime Interface](container-runtime-interface-v1.md) implements
+a standard, imperative container interface across all container runtimes
+supported by Kubernetes. This interface makes it easy to implement Image Exec.
+
+The kubelet's Generic Runtime Manager can implement `ExecInPod` with almost no
+changes to the CRI, which means we don't have to implement the logic in each of
+the container runtimes.
+
+To bind mount filesystems of other containers, the kubelet must be able to query
+the mount path of a particular container. This will be added to the `Container`
+message returned by `ListContainers`. If the container's filesystem is no longer
+mounted in the node's mount namespace then it will not be made visible to the
+troubleshooting container.
+
+`ExecInPod` will build a map of container names to node filesystem mount points
+for inclusion in the troubleshooting container's `ContainerConfig` with a
+`container_path` of `/c/<container_name>`. This can only be accomplished when
+the container is created.
+
+We could instead modify CRI's `CreateContainer` to include a list of containers
+IDs to "link" to the newly created container, but then we'd have to repeat the
+logic in each of the container backends.
+
+### Legacy Container Runtimes
+
+Implementation is slightly more difficult for the legacy docker and rkt
+runtimes.  We could implement the same create/start/attach/remove logic in rkt
+and docker_manager, or we could save effort by linking this feature to the CRI
+and return "Not Implemented" for the legacy runtimes.
+
```

Comment:
@verb Could you please add a note for hypervisor based runtime e.g.[https://github.com/hyperhq/hyperd] here or other place? Generally we think this will not break hypervisor based runtime (though may require refactoring on runtime side), but we should definitely keep them in mind.